Test rest coercion

bajtos opened the pull request:
An integration test suite describing and verifying argument coercion in REST.
 - [x] arguments in query string
 - [x]  arguments in JSON body (formdata)
 - [x] arguments in url-encoded formdata
 - [x] the JSON body as an argument

```
1230 passing (3s)
```

(see https://github.com/strongloop-internal/scrum-loopback/issues/706#issuecomment-217371058 for more information)

In this pull request, I am describing the current behaviour in the `master` branch. My intention is to run the test suite against different strong-remoting versions to find differences, I'll post results later.

> In order to discover all breaking changes and document the new behaviour, we should write an extensive integration test suite that can be run against different versions of strong-remoting. Here is an initial draft for inspiration: https://gist.github.com/bajtos/b85e53a79ba06671e08d2013acccc611 Versions to run the suite against: v2.16.0, v2.16.1, latest 2.x, master (3.0-alpha). This commit may have introduced breaking change in v2.16.1: strongloop/strong-remoting@34b0665

Connect to strongloop-internal/scrum-loopback#884
Connect to strongloop-internal/scrum-loopback#913

@ritch @STRML please take a look

bajtos commented:
See also https://github.com/strongloop-internal/scrum-loopback/issues/706#issuecomment-217371058

STRML commented:
Looks good so far, I agree with the idea of testing this more thoroughly. And it does indeed appear 34b0665 broke some backcompat, but you could argue previous behavior was a bug in the first place. Still worthwhile for cleaning up for good in 3.0.

STRML commented:
See also https://github.com/strongloop/strong-remoting/pull/265 and especially https://github.com/strongloop/strong-remoting/pull/265#issuecomment-164527886 as handling this code was a comedy of errors - somebody inside strongloop force-pushed it out without any notice, then only half the PR was re-merged and released without review, causing significant breakage as the code was incomplete (it was two commits and should not have been merged incomplete).

bajtos commented:
Done. @STRML @ritch PTAL

ritch commented on test/rest-coercion/_jsonbody.context.js:
```diff
@@ -0,0 +1,73 @@
+// Copyright IBM Corp. 2014,2016. All Rights Reserved.
```
whats with the `_` ?

bajtos replied:
I have two sets of files: helper files that does not contain any tests (names starts with underscode), and ".suite.js" files with the tests. The code loading all tests files is skipping underscore files.

I don't mind using a different approach/convention, can you propose a better solution?

ritch commented on test/rest-coercion/urlencoded-object.suite.js:
```diff
@@ -0,0 +1,94 @@
+  var ERROR_BAD_REQUEST = ctx.ERROR_BAD_REQUEST;
+  var verifyTestCases = ctx.verifyTestCases;
+
+  describe('query string - object - required', function() {
```
:clap: I love seeing this... (and the other suite with all these variable input tests)

ritch commented on test/helpers/builders.js:
```diff
@@ -0,0 +1,24 @@
+    //     Sat Jan 01 2000 00:00:00 GMT+0100 (CET)
+    // I think we should use `.toJSON()` instead, which returns concise
+    //     1999-12-31T23:00:00.000Z
+    '$data': new Date(val).toString(),
```
Yeah this should not use `toString()`. Is there a way to support both? Maybe a format setting somewhere. Not in this PR though.

bajtos replied:
Let's discuss. I can see two approaches: 

1. A global feature flag that will control whether date is serialized via `toString()` or `toJSON()`. The application author decides the response format for all clients.

2. A special request argument (or HTTP header) allowing the client to request dates in the old strong-remoting@2.x format.

Thoughts?

bajtos replied:
I have reworked the tests to not depend on this any more. Let's move the discussion to https://github.com/strongloop/loopback/issues/1990.

ritch commented on test/rest-coercion/urlencoded-date.suite.js:
```diff
@@ -0,0 +1,76 @@
+
+  describe('query string - date - optional', function() {
+    verifyTestCases({ arg: 'arg', type: 'date' }, [
+      [EMPTY_QUERY, undefined],
```
Just a nit: its hard to tell what the eventual assertions are here... a comment would be enough.

bajtos replied:
Done, see 20a40ce

bajtos commented:
I have cleaned up the code a lot, restructured test cases and added comments to make it more clear what is being tested. The patch is ready for another round of code review.

@STRML @ritch PTAL.

bajtos commented on test/rest-coercion/README.md:
```diff
@@ -0,0 +1,82 @@
+
+## URL path parameters
+
+We don't have coverage for this scenario yet.
```
I created an internal story to keep track of the task of adding test coverage for URL path parameters, see https://github.com/strongloop-internal/scrum-loopback/issues/921

bajtos commented:
The high-level overview of changes between 2.16.x versions can be found here: https://gist.github.com/bajtos/8afaa9d1ed7d7b410c76ec2095a8a992

ritch commented:
LGTM

I don't see anything that should change. But there is quite a bit I may be overlooking still. I think the size of this test suite points to a separate module making sense for coercion.

The documentation in these tests would also be helpful to everyone. I think we should create a section about coercion in the readme that lists out the nuances of coercion. Which can be done outside of this PR.

Amir-61 commented:
@slnode test please

Amir-61 commented:
I could not find any problem either. The patch LGTM.

davidcheung commented:
LGTM too
should be really helpful for us to prevent regression on a much more detailed level :+1:

bajtos commented:
@ritch @Amir-61 @davidcheung thank you for the review!

> The documentation in these tests would also be helpful to everyone. I think we should create a section about coercion in the readme that lists out the nuances of coercion. Which can be done outside of this PR.

I filled a new task for that - https://github.com/strongloop/strong-remoting/issues/313, I think it will be best to document the _new_ behaviour after we make the intended changes/fixes.
